src/dispatch/incident_cost_model/service.py

```diff
@@ -0,0 +1,215 @@
+            )
+            return False
+        seen.add(activity.plugin_event.id)
+    return True
+
+
+def get_incident_cost_model_activity_by_id(
+    *, db_session, incident_cost_model_activity_id: int
+) -> IncidentCostModelActivity:
+    """Returns an incident cost model based on the given incident cost model id."""
+    return (
+        db_session.query(IncidentCostModelActivity)
+        .filter(IncidentCostModelActivity.id == incident_cost_model_activity_id)
+        .one()
+    )
+
+
+def delete_incident_cost_model_activity(*, db_session, incident_cost_model_activity_id: int):
+    """Deletes an incident cost model activity."""
+    incident_cost_model_activity = get_incident_cost_model_activity_by_id(
+        db_session=db_session, incident_cost_model_activity_id=incident_cost_model_activity_id
+    )
+    db_session.delete(incident_cost_model_activity)
+    db_session.commit()
+
+
+def update_incident_cost_model_activity(
+    *, db_session, incident_cost_model_activity_in: IncidentCostModelActivityUpdate
+):
+    """Updates an incident cost model activity."""
+    incident_cost_model_activity = IncidentCostModelActivity(
+        response_time_seconds=incident_cost_model_activity_in.response_time_seconds,
+        enabled=incident_cost_model_activity_in.enabled,
+        plugin_event_id=incident_cost_model_activity_in.plugin_event.id,
+    )
+
+    incident_cost_model_activity.response_time_seconds = (
+        incident_cost_model_activity_in.response_time_seconds
+    )
+    incident_cost_model_activity.enabled = incident_cost_model_activity_in.enabled
+    incident_cost_model_activity.plugin_event_id = incident_cost_model_activity_in.plugin_event.id
+
+    db_session.commit()
+    return incident_cost_model_activity
+
+
+def create_incident_cost_model_activity(
+    *, db_session, incident_cost_model_activity_in: IncidentCostModelActivityCreate
+) -> IncidentCostModelActivity:
+    incident_cost_model_activity = IncidentCostModelActivity(
+        response_time_seconds=incident_cost_model_activity_in.response_time_seconds,
+        enabled=incident_cost_model_activity_in.enabled,
+        plugin_event_id=incident_cost_model_activity_in.plugin_event.id,
+    )
+
+    db_session.add(incident_cost_model_activity)
+    db_session.commit()
+    return incident_cost_model_activity
+
+
+def delete(*, db_session, incident_cost_model_id: int):
+    """Deletes an incident cost model."""
+    incident_cost_model = get_cost_model_by_id(
+        db_session=db_session, incident_cost_model_id=incident_cost_model_id
+    )
+    if not incident_cost_model:
+        raise ValueError(
+            "Unable to delete incident cost model. No incident cost model found with that id."
+        )
+
+    for activity in incident_cost_model.activities:
+        incident_cost_model_service.delete_incident_cost_model_activity(
+            db_session=db_session, incident_cost_model_activity_id=activity.id
+        )
```

Comment:
I think you might be able to avoid this loop by adding on delete cascade to the model.